Set tags from user defined field "tags" #1

xverges opened the pull request:
The field "tags" is interpreted as the tags to be set in a note. 
Tag additions/removals performed outside of the file system notes are discarded when importing again the same file

xverges committed:
Set tags from user defined field "tags" #1

amarmaduke commented:
Thanks for taking a stab at this.

A couple of problems:
- In order to check if a key exists in a dictionary you'll need syntax like `if "tags" in textfile:`
- If a tag changes on a card it doesn't update that a card was changed
- If a tag is removed (i.e. after an import no card has that tag on it) it's not cleaned up (i.e. deleted)

If you could fix up the first bullet point I can fill in the rest.

xverges committed:
Check for keys the python way, not the JS way!

xverges commented:
1. Oops!
2. I think that this is already taken care of: changing a tag changes the "tags" field, and the regular field changed branch kicks in
3. Thanks! I didn't even consider that this could be automated and thought that it needed manual db maintenance.

amarmaduke commented:
Let's do something like this for when we're potentially changing tags:

``` python
new_tags = ""
.
.
.
if "tags" in n:
   for t in n["tags"].split():
        new_tags = col.tags.addToStr(t, new_tags)
   tags_to_register.extend(col.tags.split(new_tags))
   new_tags = col.tags.addToStr("ffsi:owned", new_tags)
else:
   new_tags = note[5]
if note[5] != new_tags:
   changed = True
if changed:
   new_tags = col.tags.addToStr("ffsi:changed", new_tags)
   changedc = changedc + 1
else:
   new_tags = col.tags.remFromStr("ffsi:changed", new_tags)
new_tags = col.tags.remFromStr("ffsi:added", new_tags)
note[5] = new_tags
```

For good measure I have `new_tags = ""` in the variable initialization group (around line 140). I prefer this way of handling it because it uses the internal `col.tags.*` calls. I hadn't looked too deeply into how tags are handled in the database, but fields are separated by a special unicode character. Alas so are tags, so i'd rather we make sure the formatting in the database is correct.

This also adds a changed check, if you're changing a note in-place we wont catch it if just the tags change. The "tags" field on the file system doesn't become an actual _field_ inside anki unless you have a model that picks it up.

For the third issue, just slap this as the very last call: `col.fixIntegrity()` that'll clean up any orphan tags.

Do you mind transporting those changes here? I think we're good as far as the pull request after that. I'll fix up the README to mention this and then update on ankiweb.

xverges committed:
Check for changes in tags

Tags don't need to be in the model to be managed by the import, so we cannot rely on
a bulk check for changes in all fields to know if tags have changed

xverges commented:
Your suggested changes worked well, except that they were causing the "changed" flag to be raised always due to the "fssi:" tags. Once that was taken care of, the import worked fine. Thanks!

xverges commented:
@amarmaduke , if `[[tags]]` is not present, I am considering that all tags should be cleared. Maybe the meaning should be "I am managing my tags from the Anki GUI and ffsi should leave them alone"?

xverges committed:
If no `tags` field is present, do not delete tags

amarmaduke commented:
In general I like to make the the changes of the import not effect independent changes in GUI as much as possible. Obviously the intent is that you're buying into the concept of doing everything on the file system.

In the future i'll probably want directories to act as a tagging mechanism, I'd prefer that [[tags]] just not add additional tags (allowing ffsi:\* tags to be added, as they're used to determine ownership and other helpful things). So I wouldn't want any internal tags removed if [[tags]] wasn't present, but I wouldn't mind user-defined tags removed if ffsi:owned is present but [[tags]] is not.

xverges commented:
I pushed aaca349 before I saw your comment. It prevents the clearing of the tags when `[[tags]]` is not present
